doc/code_data_relocation/relocation.rst

```diff
@@ -0,0 +1,97 @@
+.. _code_data_relocation:
+
+CODE DATA RELOCATION
+####################
+
+Overview
+********
+This script will relocate .text, .rodata, .data and .bss sections from
+required files and places it in the required memory region. This
+memory region and file are given to this python script in the form
+of a string. This script is always invoked from inside cmake.
+
+This script provides a robust way to re-order
+the memory contents without actually having to modify the code.
+In simple terms this script will do the job of
+``__attribute__((section("name")))`` for a bunch of files together.
+
+Details
+*******
+The memory region and file are given to this python script in the form of a string.
+
+Example of such a string would be:
+``SRAM2:/home/xyz/zephyr/samples/hello_world/src/main.c,SRAM1:/home/xyz/zephyr/samples/hello_world/src/main2.c``
+
+This script is invoked with following parameters:
+``python3 gen_relocate_app.py -i input_string -o generated_linker -c generated_code``
+
+Kconfig :option:`CONFIG_CODE_DATA_RELOCATION` option,  when enabled in ``prj.conf``, will
+invoke the script and do the required relocation.
+
+This script also trigger the generation of ``linker_relocate.ld`` and ``code_relocation.c`` files.
+The ``linker_relocate.ld`` file creates appropriate sections and
+link the required functions or variables from all the selected files.
+
+**Note**:
+On running this script text section is split into 2 parts. The first section
+will have some info regarding vector tables and debug info.
+The second section will have the complete text section.
+This is needed to force the required functions and data variables the correct locations.
+This is due to the behavior of the linker. The linker will only link once and
+hence this text section had to be split to make room for the generated linker script.
+
+The ``code_relocation.c`` file has code that is needed for
+initializing data sections and a copy of text sections (if XIP).
+Also this contains code needed for bss zeroing and
+for  data copy operations from ROM to required memory type.
+
+**The procedure to invoke this feature is:**
+
+* Enable :option:`CONFIG_CODE_DATA_RELOCATION` in the ``prj.conf`` file
+
+* Inside the ``CMakeLists.txt`` file in the project. we need to mention
+  all the files that needs to get relocated.
+
+  ``zephyr_code_relocate(src/*.c SRAM2)``
+
+  Where the first argument is the file/files and the second
+  argument is the memory where it must be placed.
+
+ .. note::
+    The file argument supports limited regular expressions.
+    function zephyr_code_relocate() can be called  as many times as required.
+    This step has to be performed before the inclusion of boilerplate.cmake.
+
+
+Additional Configurations
+=========================
+This section shows additional configuration options that can be set in ``CMakeLists.txt``
+
+* if the memory is SRAM1, SRAM2, CCD, or AON, then place the full object in the sections
+  for example:
+
+.. code-block:: none
```

Comment:
This code block (and contents) should be indented by two spaces to be treated as part of the bullet list item.